src/action-menu/ActionMenu.jsx

```diff
@@ -0,0 +1,146 @@
+import React, { useRef } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import classNamesBind from 'classnames/bind';
+import ThemeContext from 'terra-theme-context';
+import ContentContainer from 'terra-content-container';
+import Header from './ActionMenuHeader';
+import {
+  generateOnKeyDown,
+  itemByDirection,
+  itemByChar,
+  flattenActionItems,
+} from './_ActionUtils';
+
+import styles from './ActionMenu.module.scss';
+
+const cx = classNamesBind.bind(styles);
+
+const propTypes = {
+  /**
+   * Label for the ActionMenu.
+   */
+  ariaLabel: PropTypes.string.isRequired,
+  /**
+   * The child ActionMenu elements.
+   */
+  children: PropTypes.node,
+  /**
+   * Whether or not a header should be displayed using the ariaLabel.
+   */
+  isHeaderDisplayed: PropTypes.bool,
+  /**
+   * The close callback.
+   */
+  onRequestClose: PropTypes.func,
+  /**
+   * @private
+   * Callback function for event.
+   */
+  isHeightBounded: PropTypes.bool,
+  /**
+   * @private
+   * Callback function for event.
+   */
+  isWidthBounded: PropTypes.bool,
+};
+
+const ActionMenu = ({
+  ariaLabel,
+  children,
+  isHeightBounded,
+  isHeaderDisplayed,
+  isWidthBounded,
+  onRequestClose,
+  ...customProps
+}) => {
+  const menuRef = useRef();
+  const { items, indentChildren } = flattenActionItems(children);
+
+  const focusItem = item => {
```

Comment:
is it worth extracting the `focusItem`, `onArrow`, and `onChar` helpers outside the component definition, and pass in the `ref` and `items` as arguments, instead of relying on `ref` and `items` being defined in the outer scope? it could help separate responsibilities